base/strings/util.jl

```diff
@@ -388,13 +388,83 @@ _replace(io, repl, str, r, pattern) = print(io, repl)
 _replace(io, repl::Function, str, r, pattern) =
     print(io, repl(SubString(str, first(r), last(r))))
 
-replace(str::String, pat_repl::Pair{Char}; count::Integer=typemax(Int)) =
-    replace(str, equalto(first(pat_repl)) => last(pat_repl); count=count)
-replace(str::String, pat_repl::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}};
-        count::Integer=typemax(Int)) =
-    replace(str, occursin(first(pat_repl)) => last(pat_repl), count)
+# replace Char with function that compares with this Char
+pairmap(p::Pair{Char}) = equalto(p.first) => p.second
+
+# replace collection of Char with function that checks occurrence in this collection
+function pairmap(p::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}})
+    occursin(p.first) => p.second
+end
+pairmap(p::Pair) = p
+
+# find the first of a list of patterns.
+# if several match at the same start position prefer the longer one
+# if also the length is equal, prefer the first in list
+# return value of the successful search and the pair
+
+function findnextall(pairs::NTuple{N,Pair}, s::AbstractString, st::Integer=1) where N
+    ps = start(pairs)
+    while !done(pairs, ps)
+        p, ps = next(pairs, ps)
+        r = findnext(p.first, s, st)
+        fr = first(r)
+        if fr > 0
+            minstart, minend, minp = fr, last(r), p
+            while !done(pairs, ps)
+                p, ps = next(pairs, ps)
+                r = findnext(p.first, s, st)
+                i, k = first(r), last(r)
+                if i > 0 && ( i < minstart || i == minstart && k > minend)
+                    minstart, minend, minp = i, k, p
+                end
+            end
+            return minstart:minend, minp
+        end
+    end
+    0:-1, (""=>"")
+end
+
+# cover the multiple pairs case
+
+function replace(str::String, pat_repls::Pair...; count::Integer=typemax(Int))
+    if count == 0 || length(pat_repls) == 0
+        return str
+    end
+    count < 0 && throw(DomainError(count, "`count` must be non-negative."))
+    pat_repls = pairmap.(pat_repls)
+    n = 1
+    e = endof(str)
+    i = a = start(str)
+    r, (pattern, repl) = findnextall(pat_repls, str, i)
+    j, k = first(r), last(r)
+    out = IOBuffer(StringVector(floor(Int, 1.2sizeof(str))), true, true)
```

Comment:
Maybe add a comment to explain the choice of this constant? Why not start with the size of the input string?

Reply:
Ok, this code is inherited. I can only speculate. Maybe the author made extensive probability studies?

Reply:
Ah, OK. Let's leave it as-is then, unless somebody can explain this choice.